File: c++/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int index = 0;
        return build(preorder, inorder, index, 0, inorder.size() - 1);
    }
private:
    TreeNode* build(vector<int>& preorder, vector<int>& inorder, int& index, int i, int j) {
        if (i > j) {
            return NULL;
        }
        
        TreeNode* root = new TreeNode(preorder[index]);
        
        int split = 0;
        for (int i = 0; i < inorder.size(); i++) {
            if (preorder[index] == inorder[i]) {
                split = i;
                break;
            }
        }
        index++;
        
        root->left = build(preorder, inorder, index, i, split - 1);
        root->right = build(preorder, inorder, index, split + 1, j);
        
        return root;
    }
};

File: kotlin/0105-construct-binary-tree-from-preorder-and-inorder-traversal.kt
class Solution {
    fun buildTree(preorder: IntArray, inorder: IntArray): TreeNode? {
        val prStart = 0; val inStart = 0
        val prEnd = preorder.size-1; val inEnd = inorder.size-1

        return helper(preorder, prStart, prEnd, inorder, inStart, inEnd);
    }

    fun helper(preorder: IntArray, prStart: Int, prEnd: Int, inorder: IntArray, inStart: Int, inEnd: Int): TreeNode? {
        if (prStart > prEnd || inStart > inEnd)
            return null

        val root = TreeNode(preorder[prStart])
        var i = 0

        while (i < inorder.size && inorder[i] != preorder[prStart]) {
            i++
        }

        root.left = helper(preorder, prStart+1, prStart-inStart+i, inorder, inStart, i-1)
        root.right = helper(preorder, prStart-inStart+i+1, prEnd, inorder, i+1, inEnd)

        return root
    }
}
